src/widgets/ThemedWrappers.h

```diff
@@ -0,0 +1,141 @@
+/*!********************************************************************
+
+   Audacity: A Digital Audio Editor
+
+   @file ThemedWrappers.h
+
+   @author Vitaly Sverchinsky
+
+   A set of classes which decorate given base classes with
+   simplified theme support. Note that classes are declared
+   final, which means they could be only used 'in-place'.
+   
+**********************************************************************/
+
+#pragma once
+
+#include <optional>
+
+#include "Internat.h"
+#include "Observer.h"
+#include "Theme.h"
+
+template<class WindowBase>
+class ThemedWindowWrapper final : public WindowBase
+{
+   Observer::Subscription mThemeChangeSubscription;
+   std::optional<TranslatableString> mTranslatableLabel;
+   int mForegroundColorIndex { -1 };
+   int mBackgroundColorIndex { -1 };
+public:
+   template <typename... Args>
+   ThemedWindowWrapper(Args&&... args) : WindowBase( std::forward<Args>(args)... )
+   {
+      mThemeChangeSubscription =
+         theTheme.Subscribe(*this, &ThemedWindowWrapper::OnThemeChange);
+   }
+
+   void SetBackgroundColorIndex(int index)
+   {
+      mBackgroundColorIndex = index;
+      if(index != -1)
+         WindowBase::SetBackgroundColour(theTheme.Colour(mBackgroundColorIndex));
+   }
+
+   void SetForegroundColorIndex(int index)
+   {
+      mForegroundColorIndex = index;
+      if(index != -1)
+         WindowBase::SetForegroundColour(theTheme.Colour(mForegroundColorIndex));
+   }
+
+   void SetTranslatableLabel(TranslatableString label)
+   {
+      mTranslatableLabel = std::move(label);
+      WindowBase::SetLabel(mTranslatableLabel->Translation());
+   }
+
+protected:
+   void OnThemeChange(ThemeChangeMessage message)
+   {
+      if (message.appearance)
+      {
+         if(mBackgroundColorIndex != -1)
+            WindowBase::SetBackgroundColour(theTheme.Colour(mBackgroundColorIndex));
+         if(mForegroundColorIndex != -1)
+            WindowBase::SetForegroundColour(theTheme.Colour(mForegroundColorIndex));
+         WindowBase::Refresh();
+      }
+      if(mTranslatableLabel)
+         WindowBase::SetLabel(mTranslatableLabel->Translation());
```

Comment:
You need to update the text of labels and buttons for changes of language -- and deriving from `PrefsListener` is how that is done.

It is not sufficient to update the labels when the theme changes -- and I don't understand why it is necessary then either, but maybe I don't know.

Ah, but is it that the dismissing of the Preferences dialog coincidentally sends theme change and PrefsListener messages together, always?

So maybe this appears to work when you test change of language, but still I don't think it is done properly.  And who knows, we might invent a different UI for language change some day that doesn't have this coincidence.

Reply:
See what was done in commit 20ace31

Reply:
>Ah, but is it that the dismissing of the Preferences dialog coincidentally sends theme change and PrefsListener messages together, always?

I didn't think it was a coincidence, as I understand making a change to a single parameter in a settings group ("interface" or "effects") updates values for all parameters in that group